target/ppc/translate/vmx-impl.c.inc

```diff
@@ -1515,6 +1515,60 @@ TRANS(VSTRIBR, do_vstri, gen_helper_VSTRIBR)
 TRANS(VSTRIHL, do_vstri, gen_helper_VSTRIHL)
 TRANS(VSTRIHR, do_vstri, gen_helper_VSTRIHR)
 
+static bool trans_VCLRLB(DisasContext *ctx, arg_VX *a)
+{
+    TCGv_i64 hi, lo, rb;
+    TCGLabel *l, *end;
+
+    REQUIRE_INSNS_FLAGS2(ctx, ISA310);
+    REQUIRE_VECTOR(ctx);
+
+    l = gen_new_label();
+    end = gen_new_label();
+
+    hi = tcg_const_local_i64(0);
+    lo = tcg_const_local_i64(0);
+    rb = tcg_temp_local_new_i64();
+
+    tcg_gen_extu_tl_i64(rb, cpu_gpr[a->vrb]);
+
+    /* RB == 0: all zeros */
+    tcg_gen_brcondi_i64(TCG_COND_EQ, rb, 0, end);
```

Comment:
You can probably reorganize these brcond's so that you don't have to read from a->vra unnecessarily in some cases:

```
    (...)

    /* RB == 0: all zeros */
    tcg_gen_brcondi_i64(TCG_COND_EQ, rb, 0, end);

    get_avr64(lo, a->vra, false);

    /* RB <= 8 */
    tcg_gen_brcondi_i64(TCG_COND_LEU, rb, 8, l);

    get_avr64(hi, a->vra, true);

    /* RB >= 16: just copy VRA to VRB */
    tcg_gen_brcondi_i64(TCG_COND_GEU, rb, 16, end);

    (continue to code that deals with 8 < rb < 16)
```

And you can also remove ``tcg_gen_movi_i64(hi, 0);`` from the RB < 8 block

Reply:
I did it this way to allow the sharing of the two calls to `set_avr64` after the `end` label. That way we have one or two fewer instructions, and it felt more readable, but I can change it if you prefer the `set_avr64`s before the `tcg_gen_br(end) `s.

Reply:
Maybe I'm missing something, but what I suggested wouldn't change that. Here's the full code:

```
    hi = tcg_const_local_i64(0);
    lo = tcg_const_local_i64(0);
    rb = tcg_temp_local_new_i64();

    tcg_gen_extu_tl_i64(rb, cpu_gpr[a->vrb]);

    /* RB == 0: all zeros */
    tcg_gen_brcondi_i64(TCG_COND_EQ, rb, 0, end);

    get_avr64(lo, a->vra, false);

    /* RB <= 8 */
    tcg_gen_brcondi_i64(TCG_COND_LEU, rb, 8, l);

    get_avr64(hi, a->vra, true);

    /* RB >= 16: just copy VRA to VRB */
    tcg_gen_brcondi_i64(TCG_COND_GEU, rb, 16, end);

    /* 8 < RB < 16: copy lo and partially clear hi */
    tcg_gen_subfi_i64(rb, 16, rb);
    tcg_gen_shli_i64(rb, rb, 3);
    tcg_gen_shl_i64(hi, hi, rb);
    tcg_gen_shr_i64(hi, hi, rb);
    tcg_gen_br(end);

    /* 0 < RB <= 8: zeroes hi and partially clears lo */
    gen_set_label(l);
    tcg_gen_subfi_i64(rb, 8, rb);
    tcg_gen_shli_i64(rb, rb, 3);
    tcg_gen_shl_i64(lo, lo, rb);
    tcg_gen_shr_i64(lo, lo, rb);

    /* Update VRT */
    gen_set_label(end);
    set_avr64(a->vrt, hi, true);
    set_avr64(a->vrt, lo, false);

    tcg_temp_free_i64(hi);
    tcg_temp_free_i64(lo);
    tcg_temp_free_i64(rb);
```

Reply:
You are right. Fixed the next commit with edf7886f6ab436e7fe7db35b5a9a7a3dedb228d3 to avoid force push. I'll fix in this commit when preparing the patch series.